### PR TITLE
Add tesseract_python and tesseract_viewer_python

### PR DESCRIPTION
**Pull request by johnwason**
This pull request adds tesseract_python and tesseract_viewer_python to the main repository. See #87

### CODE REVIEW

**Comment by Levi-Armstrong** [comment]
I am going to work on setting up a docker improve the CI builds. I will add your version of swig.

**Comment by johnwason** [comment]
@Levi-Armstrong I added swig to tesseract_ext using a cmake external project and it seems to work ok.

I am stumped why the unit tests are failing. The build and test work fine locally. It looks like a PYTHONPATH error, but I am not sure how to diagnose it.

**Comment by Levi-Armstrong** [comment]
I ran into similar issues on the melodic build where the machine was running out of resources. I would try reducing the number of projects an jobs similar to the melodic build.

**Comment by johnwason** [comment]
I ran it locally on my machine with 64 GB of ram using the following instructions:

https://github.com/ros-industrial/industrial_ci/blob/legacy/doc/index.rst#id47

and saw the same error.

**Comment by Levi-Armstrong** [comment]
What error are you getting?

**Comment by johnwason** [comment]
I am seeing "module "tesseract" not found" when running the unit tests. This suggests there is an issue with PYTHONPATH during industrial_ci that is not occurring when building normally. I tried using catkin tools instead of catkin_make_isolated, and it worked in my development space.

**Comment by Levi-Armstrong** [comment]
@johnwason I updated Travis CI config file to use a docker image to speed up the CI build and I include the swig version you had in your Add external package. We just merged the PR. Let me know if this solves the issue you were having or if I need to make modification to the docker. In theory you should be able to pull down the docker and test building it in the docker on your machine and if it builds there it should build in CI.

**Comment by Levi-Armstrong** [comment]
I am going to work on rebasing this on the latest, but may have questions.

**Comment by johnwason** [comment]
I will take a look tomorrow as well. If all you changed were some function signatures it shouldn't be too difficult to update.

**Comment by johnwason** [comment]
@Levi-Armstrong I got this error when building using travis.yml

    CMake Error at /usr/share/cmake-3.10/Modules/FindPackageHandleStandardArgs.cmake:137 
    (message):
      Could NOT find SWIG: Found unsuitable version "3.0.12", but required is at
      least "4.0.0" (found /usr/bin/swig3.0)

**Comment by Levi-Armstrong** [comment]
I will check the docker.

**Comment by Levi-Armstrong** [comment]
I looks like it is in the path
Environment:
```
root@swri:/# env
LC_ALL=C.UTF-8
LS_COLORS=rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:ow=34;42:st=37;44:ex=01;32:*.tar=01;31:*.tgz=01;31:*.arc=01;31:*.arj=01;31:*.taz=01;31:*.lha=01;31:*.lz4=01;31:*.lzh=01;31:*.lzma=01;31:*.tlz=01;31:*.txz=01;31:*.tzo=01;31:*.t7z=01;31:*.zip=01;31:*.z=01;31:*.Z=01;31:*.dz=01;31:*.gz=01;31:*.lrz=01;31:*.lz=01;31:*.lzo=01;31:*.xz=01;31:*.zst=01;31:*.tzst=01;31:*.bz2=01;31:*.bz=01;31:*.tbz=01;31:*.tbz2=01;31:*.tz=01;31:*.deb=01;31:*.rpm=01;31:*.jar=01;31:*.war=01;31:*.ear=01;31:*.sar=01;31:*.rar=01;31:*.alz=01;31:*.ace=01;31:*.zoo=01;31:*.cpio=01;31:*.7z=01;31:*.rz=01;31:*.cab=01;31:*.wim=01;31:*.swm=01;31:*.dwm=01;31:*.esd=01;31:*.jpg=01;35:*.jpeg=01;35:*.mjpg=01;35:*.mjpeg=01;35:*.gif=01;35:*.bmp=01;35:*.pbm=01;35:*.pgm=01;35:*.ppm=01;35:*.tga=01;35:*.xbm=01;35:*.xpm=01;35:*.tif=01;35:*.tiff=01;35:*.png=01;35:*.svg=01;35:*.svgz=01;35:*.mng=01;35:*.pcx=01;35:*.mov=01;35:*.mpg=01;35:*.mpeg=01;35:*.m2v=01;35:*.mkv=01;35:*.webm=01;35:*.ogm=01;35:*.mp4=01;35:*.m4v=01;35:*.mp4v=01;35:*.vob=01;35:*.qt=01;35:*.nuv=01;35:*.wmv=01;35:*.asf=01;35:*.rm=01;35:*.rmvb=01;35:*.flc=01;35:*.avi=01;35:*.fli=01;35:*.flv=01;35:*.gl=01;35:*.dl=01;35:*.xcf=01;35:*.xwd=01;35:*.yuv=01;35:*.cgm=01;35:*.emf=01;35:*.ogv=01;35:*.ogx=01;35:*.aac=00;36:*.au=00;36:*.flac=00;36:*.m4a=00;36:*.mid=00;36:*.midi=00;36:*.mka=00;36:*.mp3=00;36:*.mpc=00;36:*.ogg=00;36:*.ra=00;36:*.wav=00;36:*.oga=00;36:*.opus=00;36:*.spx=00;36:*.xspf=00;36:
LD_LIBRARY_PATH=/opt/ros/melodic/lib
LANG=C.UTF-8
HOSTNAME=swri
ROS_ETC_DIR=/opt/ros/melodic/etc/ros
PWD=/
HOME=/root
CMAKE_PREFIX_PATH=/opt/ros/melodic
ROS_ROOT=/opt/ros/melodic/share/ros
ROS_MASTER_URI=http://localhost:11311
ROS_VERSION=1
TERM=xterm
ROS_PYTHON_VERSION=2
SHLVL=1
PYTHONPATH=/opt/ros/melodic/lib/python2.7/dist-packages
ROS_PACKAGE_PATH=/opt/ros/melodic/share
ROSLISP_PACKAGE_DIRECTORIES=
PATH=/opt/ros/melodic/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
PKG_CONFIG_PATH=/opt/ros/melodic/lib/pkgconfig
ROS_DISTRO=melodic
_=/usr/bin/env

```

Version:

```
root@swri:/usr/local/bin# swig -version

SWIG Version 4.0.1

Compiled with g++ [x86_64-pc-linux-gnu]

Configured options: +pcre

Please see http://www.swig.org for reporting bugs and further information
```

File: tesseract_python/tesseract_python/CMakeLists.txt [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+  tesseract_add_run_tests_target()
+
+  add_test (NAME python-tests
+  COMMAND ${PYTHON_EXECUTABLE} -m pytest -p no:cacheprovider
+  WORKING_DIRECTORY ${CMAKE_CURRENT_SOURCE_DIR}
+  )
+  
+  find_package(tesseract_support REQUIRED)
+  if (WIN32)
+  set_tests_properties(python-tests PROPERTIES ENVIRONMENT "TESSERACT_SUPPORT_DIR=${TESSERACT_SUPPORT_DIR}")
```
Comment (Levi-Armstrong) [inline tesseract_python/tesseract_python/CMakeLists.txt]: Could this be extracted from the target INTERFACE_COMPILE_DEFINITIONS?

```
set_target_properties(tesseract::tesseract_support PROPERTIES
  INTERFACE_COMPILE_DEFINITIONS "TESSERACT_SUPPORT_DIR=\"/home/larmstrong/catkin_ws/trajopt_only_ws/install/share/tesseract_support\""
)
```

Comment (johnwason) [reply tesseract_python/tesseract_python/CMakeLists.txt]: Possibly, but it would require parsing INTERFACE_COMPILE_DEFINITIONS to extract that value

**Comment by johnwason** [comment]
@Levi-Armstrong it should have found swig with those settings. I will try `-DSWIG_EXECUTABLE` and see if it fixes the problem.

**Comment by johnwason** [comment]
This built successfully for kinetic, but timed out for melodic on my fork.

**Comment by johnwason** [comment]
Hooray! This should be ready to merge.

**Comment by Levi-Armstrong** [comment]
@johnwason We plan to separate the tesseract_ros packages into its own repository to better support [tesseract_ros2](https://github.com/ros-industrial-consortium/tesseract_ros2). Will it be difficult to separate the python wrappers?

**Comment by johnwason** [comment]
Having a separate repo should be fine. Create a tesseract_python repo and give me maintainer access.

File: tesseract_python/tesseract_python/swig/tesseract/tesseract.i [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+  using Ptr = std::shared_ptr<Tesseract>;
+  using ConstPtr = std::shared_ptr<const Tesseract>;
+
+  Tesseract();
+  virtual ~Tesseract() = default;
+
+  bool isInitialized() const;
+
+  bool init(tesseract_scene_graph::SceneGraph::Ptr scene_graph);
+  //bool init(tesseract_scene_graph::SceneGraph::Ptr scene_graph, tesseract_scene_graph::SRDFModel::ConstPtr srdf_model);
```
Comment (Levi-Armstrong) [inline tesseract_python/tesseract_python/swig/tesseract/tesseract.i]: Does this mean this method is not available in python? If not, what needs to change in Tesseract?

File: tesseract_python/tesseract_python/swig/tesseract/tesseract.i [hunk 3]
```diff
@@ -0,0 +1,70 @@
...
+  bool init(const std::string& urdf_string, const tesseract_scene_graph::ResourceLocator::Ptr& locator);
+  bool init(const std::string& urdf_string,
+            const std::string& srdf_string,
+            const tesseract_scene_graph::ResourceLocator::Ptr& locator);
+  //bool init(const boost::filesystem::path& urdf_path, tesseract_scene_graph::ResourceLocator::Ptr locator);
```
Comment (Levi-Armstrong) [inline tesseract_python/tesseract_python/swig/tesseract/tesseract.i]: Same as above.

File: tesseract_python/tesseract_python/swig/tesseract_collision/continuous_contact_manager.i [hunk 4]
```diff
@@ -0,0 +1,99 @@
...
+  virtual double getContactDistanceThreshold() const = 0;
+
+  // TODO: IsContactAllowedFn
+  //virtual void setIsContactAllowedFn(IsContactAllowedFn fn) = 0;
+
+  //virtual IsContactAllowedFn getIsContactAllowedFn() const = 0;
+  
+  //virtual void contactTest(ContactResultMap& collisions, const ContactTestType& type) = 0;
+  %extend {
+    tesseract_collision::ContactResultMap contactTest(const tesseract_collision::ContactTestType& type)
```
Comment (Levi-Armstrong) [inline tesseract_python/tesseract_python/swig/tesseract_collision/continuous_contact_manager.i]: Why was this needed?

File: tesseract_python/tesseract_python/swig/tesseract_collision/types.i [hunk 5]
```diff
@@ -0,0 +1,201 @@
...
+{
+    tesseract_collision::ContactResultVector v;
+    tesseract_collision::flattenResults(std::move(m),v);
+    return v;
+}
+}
+
+namespace tesseract_collision
+{
+/*struct ContactTestData
```
Comment (Levi-Armstrong) [inline tesseract_python/tesseract_python/swig/tesseract_collision/types.i]: May not want to expose this. This should only be used internally by the collision libraries.

File: tesseract_python/tesseract_python/swig/tesseract_common/types.i [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+tesseract_aligned_map(name, %arg(Key), %arg(std::vector<Value , Eigen::aligned_allocator<Value >>));
+%enddef
+
+%define tesseract_aligned_unordered_map(name,Key,Value)
+%template(name) std::unordered_map<Key,Value,std::hash<Key>,std::equal_to<Key>,Eigen::aligned_allocator<std::pair<const Key, Value>>>;
+%enddef
+
+tesseract_aligned_vector(VectorIsometry3d, Eigen::Isometry3d);
+//tesseract_aligned_vector(VectorVector3d, Eigen::Vector3d);
+//TODO: Why is this not aligned??
```
Comment (Levi-Armstrong) [inline tesseract_python/tesseract_python/swig/tesseract_common/types.i]: This was pointed out to me by the FCL maintainers, that the Eigen::Vector3d is not a fixed-size vectorizable type based on the [documentation](https://eigen.tuxfamily.org/dox/group__TopicFixedSizeVectorizable.html).

File: tesseract_python/tesseract_python/swig/tesseract_kinematics/core/forward_kinematics.i [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+  using Ptr = std::shared_ptr<ForwardKinematics>;
+  using ConstPtr = std::shared_ptr<const ForwardKinematics>;
+
+  virtual ~ForwardKinematics() = default;
+
+%extend {
+
+  //virtual bool calcFwdKin(Eigen::Isometry3d& pose, const Eigen::Ref<const Eigen::VectorXd>& joint_angles) const = 0;
+  
+  Eigen::Isometry3d calcFwdKin(const Eigen::Ref<const Eigen::VectorXd>& joint_angles) const
```
Comment (Levi-Armstrong) [inline tesseract_python/tesseract_python/swig/tesseract_kinematics/core/forward_kinematics.i]: Why change the behavior?

File: tesseract_python/tesseract_python/swig/tesseract_planning/tesseract_motion_planners/core/planner.i [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+  virtual ~MotionPlanner();
+  
+  const std::string& getName();
+  
+  const PlannerRequest& getRequest() const;
+  
+  void setRequest(const PlannerRequest& request);
+  
+  %rename(_solve) solve;
+  virtual tesseract_common::StatusCode solve(PlannerResponse& res, bool verbose = false) = 0;
```
Comment (Levi-Armstrong) [inline tesseract_python/tesseract_python/swig/tesseract_planning/tesseract_motion_planners/core/planner.i]: Why change the behavior?

File: tesseract_python/tesseract_python/swig/tesseract_planning/tesseract_motion_planners/trajopt/trajopt_motion_planner.i [hunk 9]
```diff
@@ -0,0 +1,84 @@
...
+class TrajOptMotionPlanner : public MotionPlanner
+{
+public:
+  
+  TrajOptMotionPlanner(std::string name = "TRAJOPT");
+
+  ~TrajOptMotionPlanner() {}
+
+  //bool setConfiguration(const TrajOptPlannerConfig::Ptr config);
+  %extend {
```
Comment (Levi-Armstrong) [inline tesseract_python/tesseract_python/swig/tesseract_planning/tesseract_motion_planners/trajopt/trajopt_motion_planner.i]: Why the change?

File: tesseract_python/tesseract_python/swig/tesseract_scene_graph/allowed_collision_matrix.i [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+  using AllowedCollisionEntries = std::unordered_map<LinkNamesPair, std::string, PairHash>;
+
+  AllowedCollisionMatrix();
+  virtual ~AllowedCollisionMatrix();
+
+  virtual void addAllowedCollision(const std::string& link_name1,
+                                   const std::string& link_name2,
+                                   const std::string& reason);
+
+%extend {
```
Comment (Levi-Armstrong) [inline tesseract_python/tesseract_python/swig/tesseract_scene_graph/allowed_collision_matrix.i]: Why was this required?

**Comment by Levi-Armstrong** [comment]
> Having a separate repo should be fine. Create a tesseract_python repo and give me maintainer access.

Sorry, I meant that we will be moving the package under tesseract_ros into its own repository `tesseract_ros` so the tesseract_python will need to be split up into tesseract_python and tesseract_ros_python where tesseract_python would stay in this repository and tesseract_ros_python would go in tesseract_ros repository once it gets created.

**Comment by johnwason** [comment]
Yeah, it should be fine to split up the python into two separate repos. The only wrinkle is that the tesseract_ros_python package will need access to the *.i files of tesseract_python. They will have to be installed somewhere so that they are available later. I will look for the convention on where they should go.

File: tesseract_python/tesseract_python/swig/geometry_loaders.h [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * @par
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <tesseract_geometry/mesh_parser.h>
```
[the remark this answers is not in the document]
Comment (johnwason) [reply tesseract_python/tesseract_python/swig/geometry_loaders.h]: If you think they are useful elsewhere, they can be easily moved into the main packages. They were required to make the Python wrappers work.

File: tesseract_python/tesseract_python/swig/tesseract/tesseract.i (hunk 2 above)
Comment (johnwason) [reply tesseract_python/tesseract_python/swig/tesseract/tesseract.i]: I wasn't able to wrap the SRDF classes due to nested classes. It is also inflexible to adding and removing groups. I left the function in as a stopgap until something more polished was ready.

File: tesseract_python/tesseract_python/swig/tesseract/tesseract.i (hunk 3 above)
Comment (johnwason) [reply tesseract_python/tesseract_python/swig/tesseract/tesseract.i]: For my purposes I didn't need the file loading, so I disabled it. It can be enabled in the future if needed.

Comment (Levi-Armstrong) [reply tesseract_python/tesseract_python/swig/tesseract/tesseract.i]: Ok

File: tesseract_python/tesseract_python/swig/tesseract/tesseract.i (hunk 2 above)
Comment (Levi-Armstrong) [reply tesseract_python/tesseract_python/swig/tesseract/tesseract.i]: Makes sense. This is one of the next classes on my list to rewrite.

File: tesseract_python/tesseract_python/swig/tesseract_collision/types.i (hunk 5 above)
Comment (johnwason) [reply tesseract_python/tesseract_python/swig/tesseract_collision/types.i]: We need access to the collision checker from Python. The contact monitor example required access to these classes. Is there a specific type you don't want exposed?

File: tesseract_python/tesseract_python/swig/tesseract_common/types.i (hunk 6 above)
Comment (johnwason) [reply tesseract_python/tesseract_python/swig/tesseract_common/types.i]: I see

File: tesseract_python/tesseract_python/swig/tesseract_kinematics/core/forward_kinematics.i (hunk 7 above)
Comment (johnwason) [reply tesseract_python/tesseract_python/swig/tesseract_kinematics/core/forward_kinematics.i]: Python does not handle output parameters well. It would need to be converted to a tuple, or throw an exception on failure. The convention of Python is to throw an exception on error since exceptions are not very expensive in Python. I am probably going to change all the return codes to exceptions in the future, but I haven't had a chance to do so yet.

File: tesseract_python/tesseract_python/swig/tesseract_scene_graph/allowed_collision_matrix.i (hunk 10 above)
Comment (johnwason) [reply tesseract_python/tesseract_python/swig/tesseract_scene_graph/allowed_collision_matrix.i]: I forget the exact reason I changed this. I was probably having trouble wrapping the return type.

File: tesseract_python/tesseract_python/swig/tesseract_planning/tesseract_motion_planners/trajopt/trajopt_motion_planner.i (hunk 9 above)
Comment (johnwason) [reply tesseract_python/tesseract_python/swig/tesseract_planning/tesseract_motion_planners/trajopt/trajopt_motion_planner.i]: To match the Python convention of throwing exceptions on failure.

File: tesseract_python/tesseract_python/swig/tesseract_planning/tesseract_motion_planners/core/planner.i (hunk 8 above)
Comment (johnwason) [reply tesseract_python/tesseract_python/swig/tesseract_planning/tesseract_motion_planners/core/planner.i]: Python does not have out parameters.

**Comment by johnwason** [comment]
For the various %extend directives, I am going to see if I can use typemaps to do the change for parameter returns and return value handling. This will make the code cleaner and also automatically do the right thing with various function signatures.

**Comment by johnwason** [comment]
> Also what is the best way to maintained this going forward to catch when things change in the c++ things get updated in the python wrapper? Will Swig complain if a function exist in C++ but not in the wrapper or vice-versa?

Swig does not verify anything about the C++ files, it is just a wrapper generator. If new functions are added, it won't do anything different. If C++ functions are removed or the signature changes, the compilation will fail.

File: tesseract_python/tesseract_python/swig/tesseract_collision/types.i (hunk 5 above)
Comment (Levi-Armstrong) [reply tesseract_python/tesseract_python/swig/tesseract_collision/types.i]: Just the ContactTestData. Everything else should be good.

File: tesseract_python/tesseract_python/swig/tesseract_kinematics/core/forward_kinematics.i (hunk 7 above)
Comment (Levi-Armstrong) [reply tesseract_python/tesseract_python/swig/tesseract_kinematics/core/forward_kinematics.i]: Thanks for the explanation.

File: tesseract_python/tesseract_python/swig/tesseract_planning/tesseract_motion_planners/core/planner.i (hunk 8 above)
Comment (Levi-Armstrong) [reply tesseract_python/tesseract_python/swig/tesseract_planning/tesseract_motion_planners/core/planner.i]: I see.

File: tesseract_python/tesseract_python/swig/geometry_loaders.h (hunk 11 above)
[the remark this answers is not in the document]
Comment (Levi-Armstrong) [reply tesseract_python/tesseract_python/swig/geometry_loaders.h]: I think I could use this in the tesseract_rviz.

**Comment by johnwason** [comment]
I updated to the changes in master and the discussion here.

File: tesseract_python/tesseract_python/CMakeLists.txt [hunk 12]
```diff
@@ -0,0 +1,191 @@
+cmake_minimum_required(VERSION 2.8.3)
+project(tesseract_python)
+
+# Read version
+FILE (STRINGS "${CMAKE_CURRENT_SOURCE_DIR}/package.xml" tesseract_python_version1 REGEX "<version>[0-9]+\\.[0-9]+\\.[0-9]+</version>")
+IF ("${tesseract_python_version1}" STREQUAL "" )
+MESSAGE(FATAL_ERROR "Could not read tesseract_python version.")
+ENDIF()
+STRING (REGEX REPLACE "<version>([0-9]+\\.[0-9]+\\.[0-9]+)</version>" "\\1" tesseract_python_version ${tesseract_python_version1})
+MESSAGE(STATUS "tesseract_python version: ${tesseract_python_version}")
```
Comment (Levi-Armstrong) [inline tesseract_python/tesseract_python/CMakeLists.txt]: This looks useful so you only have to update the version in the package.xml. What are your thoughts on using this for every package? I am not asking to make the change now, but if you agree I will create a PR after this is merged adding this to tesseract_common cmake macro file with a function called extract_version.

**Comment by Levi-Armstrong** [comment]
Looks good. It looks like there are still some conflict.

File: tesseract_python/tesseract_python/CMakeLists.txt (hunk 12 above)
Comment (johnwason) [reply tesseract_python/tesseract_python/CMakeLists.txt]: Where do you need version information in the rest of Tesseract? The version information here is mainly for use with building Python wheels. For the ROS packages you probably only need the package.xml version.

I based this version extraction of this code in Robot Raconteur: https://github.com/robotraconteur/robotraconteur/blob/2ae3738e937e72c7b8096914d29e3bbeb379b71f/CMakeLists.txt#L10-L17

On a related note, it would be useful to be able to build the Python module with static dependencies so there is only the one *.so or *.pyd binary file in the Python wheel. It would make distribution a lot easier. That can be done after the pull request is accepted.

**Comment by johnwason** [comment]
Which conflicts are you seeing? I pulled master today so it should be up to date.

**Comment by Levi-Armstrong** [comment]
It looks like I can do a squash and merge. Do you see any reason not to do a squash merge?

**Comment by johnwason** [comment]
Squash and merge should be fine.

**Comment by johnwason** [comment]
You may want to add swig 4.0 to tesseract_ext

**Comment by johnwason** [comment]
Hmm, maybe I should do the squash. I will force push to the branch tomorrow with a single commit.

File: tesseract_python/tesseract_python/CMakeLists.txt (hunk 12 above)
Comment (gavanderhoorn) [reply tesseract_python/tesseract_python/CMakeLists.txt]: It would actually not be a bad idea to set the `VERSION` argument to `project(..)`, as it would then generate version-specific artefacts. If we forward the version to libraries as well, that would help prevent ABI compatibility problems because CMake would set `soversions` if we ask it to (which helps explicitly error out when users don't rebuild their workspace properly after updating packages).

In another project I've used this set of statements, which is a bit more concise:

```cmake
# Read version from the package.xml file.
file(READ ${CMAKE_CURRENT_SOURCE_DIR}/package.xml package_xml_str)
if(NOT package_xml_str MATCHES "<version>([0-9]+.[0-9]+.[0-9]+)</version>")
  message(FATAL_ERROR "Could not parse project version from package.xml. Aborting.")
endif()

# At this point we either have a proper version string, or we've errored
# out with a FATAL_ERROR above. So assume CMAKE_MATCH_1 contains our
# package's version.
project(... VERSION ${CMAKE_MATCH_1} ...)
```

All plain CMake packages get their versions from the package manifest in this way. Then use `set_target_properties(... VERSION ${${PROJECT_NAME}_VERSION ...)` on libraries and other targets where it makes sense.

**Comment by johnwason** [comment]
I squashed the commits. This should be ready to merge assuming the CI passes.

File: tesseract_python/tesseract_python/CMakeLists.txt (hunk 12 above)
Comment (Levi-Armstrong) [reply tesseract_python/tesseract_python/CMakeLists.txt]: @gavanderhoorn I will added to all of the packages! :)

**Comment by Levi-Armstrong** [comment]
@johnwason Thanks I will wait on CI.
